src/inc/til/rle.h

```diff
@@ -0,0 +1,1066 @@
+            // If you use this as a sample, all you have to change is:
+            // 1. Make it inherit correctly and align that with the template.
+            // 2. Fix mybase to match
+            // 3. value_type needs to be whatever makes sense to come off of *iterator
+            // 4. difference_type needs to come from somewhere else, probably.
+
+            using mybase = rle_const_iterator<ParentIt>;
+
+            using iterator_category = std::random_access_iterator_tag;
+            using value_type = typename ParentIt::value_type::first_type;
+            using pointer = value_type*;
+            using reference = value_type&;
+            using difference_type = typename ParentIt::difference_type;
+
+            // Use base's constructor.
+            using mybase::mybase;
+
+            [[nodiscard]] reference operator*() const noexcept
+            {
+                return const_cast<reference>(mybase::operator*());
+            }
+
+            [[nodiscard]] pointer operator->() const noexcept
+            {
+                return const_cast<std::remove_const_t<value_type>*>(mybase::operator->());
+            }
+
+            rle_iterator& operator++() noexcept
+            {
+                mybase::operator++();
+                return *this;
+            }
+
+            rle_iterator operator++(int) noexcept
+            {
+                rle_iterator tmp = *this;
+                mybase::operator++();
+                return tmp;
+            }
+
+            rle_iterator& operator--() noexcept
+            {
+                mybase::operator--();
+                return *this;
+            }
+
+            rle_iterator operator--(int) noexcept
+            {
+                rle_iterator tmp = *this;
+                mybase::operator--();
+                return tmp;
+            }
+
+            rle_iterator& operator+=(const difference_type offset) noexcept
+            {
+                mybase::operator+=(offset);
+                return *this;
+            }
+
+            [[nodiscard]] rle_iterator operator+(const difference_type offset) const noexcept
+            {
+                rle_iterator tmp = *this;
+                return tmp += offset;
+            }
+
+            rle_iterator& operator-=(const difference_type offset) noexcept
+            {
+                mybase::operator-=(offset);
+                return *this;
+            }
+
+            // Use base's difference method.
+            using mybase::operator-;
+
+            [[nodiscard]] rle_iterator operator-(const difference_type offset) const noexcept
+            {
+                rle_iterator tmp = *this;
+                return tmp -= offset;
+            }
+
+            [[nodiscard]] reference operator[](const difference_type offset) const noexcept
+            {
+                return const_cast<reference>(mybase::operator[](offset));
+            }
+        };
+    };
+
+    template<typename T, typename S = size_t>

```

Comment:
I know what would be helpful: C++20 & `std::is_integral`.

Reply:
Yeah, I know, right? I can't wait. Until then, I think @DHowett and I agreed that we can just use it correctly and let it barf gloriously in the compiler should you use the template wrong. Unless I misunderstood that conversation or he gets a new idea in the next week to scope it down.